bootstrapping/bootstrapping.go

```diff
@@ -0,0 +1,428 @@
+)
+
+type bootstrappingStatus struct {
+	startedAt      string
+	finishedAt     string
+	policiesStatus []policyStatus
+	attributes     attributesStatus
+}
+type attributesStatus struct {
+	revisionID string
+	filename   string
+	filePath   string
+	rawData    *json.RawMessage
+}
+
+type policyStatus struct {
+	id          string
+	revisionID  string
+	name        string
+	filename    string
+	tarballURL  string
+	queryURL    string
+	tarballPath string
+	folderPath  string
+
+	downloaded   bool
+	uncompressed bool
+	executed     bool
+	logged       bool
+}
+
+// Handle signals
+func handleSysSignals(cancelFunc context.CancelFunc) {
+	log.Debug("handleSysSignals")
+
+	gracefulStop := make(chan os.Signal, 1)
+	signal.Notify(gracefulStop, syscall.SIGTERM, syscall.SIGINT, syscall.SIGKILL)
+	log.Debug("Ending, signal detected:", <-gracefulStop)
+	cancelFunc()
+}
+
+// Returns the full path to the tmp folder joined with pid management file name
+func getProcessIDFilePath() string {
+	return strings.Join([]string{os.TempDir(), string(os.PathSeparator), ProcessIDFile}, "")
+}
+
+// Returns the full path to the tmp folder
+func getProcessingFolderFilePath() string {
+	dir := strings.Join([]string{os.TempDir(), string(os.PathSeparator), "imco", string(os.PathSeparator)}, "")
+	os.Mkdir(dir, 0777)
+	return dir
+}
+
+// Start the bootstrapping process
+func start(c *cli.Context) error {
+	log.Debug("start")
+
+	formatter := format.GetFormatter()
+	if err := utils.SetProcessIdToFile(getProcessIDFilePath()); err != nil {
+		formatter.PrintFatal("cannot create the pid file", err)
+	}
+
+	timingInterval := c.Int64("time")
+	if !(timingInterval > 0) {
+		timingInterval = DefaultTimingInterval
+	}
+	// Adds a random value to the given timing interval!
+	// Sleep for a configured amount of time plus a random amount of time (10 minutes plus 0 to 5 minutes, for instance)
+	timingInterval = timingInterval + int64(rand.New(rand.NewSource(time.Now().UnixNano())).Intn(DefaultRandomMaxThreshold)*60)
+	log.Debug("time interval:", timingInterval)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	go handleSysSignals(cancel)
+
+	bootstrappingRoutine(ctx, c, timingInterval)
+
+	return nil
+}
+
+// Stop the bootstrapping process
+func stop(c *cli.Context) error {
+	log.Debug("cmdStop")
+
+	formatter := format.GetFormatter()
+	if err := utils.StopProcess(getProcessIDFilePath()); err != nil {
+		formatter.PrintFatal("cannot stop the bootstrapping process", err)
+	}
+
+	log.Info("Bootstrapping routine successfully stopped")
+	return nil
+}
+
+// Main bootstrapping background routine
+func bootstrappingRoutine(ctx context.Context, c *cli.Context, timingInterval int64) {
+	log.Debug("bootstrappingRoutine")
+
+	//formatter := format.GetFormatter()
+	bootstrappingSvc, formatter := cmd.WireUpBootstrapping(c)
+	commandProcessed := make(chan bool, 1)
+
+	// initialization
+	currentTicker := time.NewTicker(time.Duration(timingInterval) * time.Second)
```

Comment:
I  theory, you should
```
defer currentTicker.Stop()
```
though in practice there is going to be just one of these tickers and once this function has exited the process will as well, so...

Reply:
Or actually, I don't know that we can operate with tickers that way.... It will not be the usual scenario, but a chef-client run could take longer than 10 minutes (if the user had a policyfile that includes some heavy computation in the run list). In that case we could be trying to run a second chef-client command while the first has not finished...

Instead we should run the chef client command, and once it has finished start a ticker that should be stopped before running the command.

Reply:
i think this can be managed thanks to commandProcessed / isRunningCommandRoutine

Reply:
But a ticker does not wait until you consume a tick to start counting for the other tick. With the following code
```Go
package main

import (
        "fmt"
        "time"
)

func main() {
        ticker := time.NewTicker(time.Second)
        defer ticker.Stop()
        for {
                t := <-ticker.C
                fmt.Printf("Current time: %v; Obtained time: %v\n", time.Now(), t)
                time.Sleep(4 * time.Second)
                fmt.Printf("Current time: %v; waiting for ticker\n", time.Now())
        }
}
```

I obtain the following output:
```
Current time: 2019-03-05 16:14:45.625772413 +0100 CET m=+1.000373507; Obtained time: 2019-03-05 16:14:45.625723294 +0100 CET m=+1.000324338
Current time: 2019-03-05 16:14:49.626468462 +0100 CET m=+5.001069656; waiting for ticker
Current time: 2019-03-05 16:14:49.626543306 +0100 CET m=+5.001144234; Obtained time: 2019-03-05 16:14:46.625735049 +0100 CET m=+2.000336092
Current time: 2019-03-05 16:14:53.626677706 +0100 CET m=+9.001278735; waiting for ticker
Current time: 2019-03-05 16:14:53.626798406 +0100 CET m=+9.001399352; Obtained time: 2019-03-05 16:14:50.625681502 +0100 CET m=+6.000282430
Current time: 2019-03-05 16:14:57.626931413 +0100 CET m=+13.001532273; waiting for ticker
Current time: 2019-03-05 16:14:57.626987325 +0100 CET m=+13.001588147; Obtained time: 2019-03-05 16:14:54.625680754 +0100 CET m=+10.000281625
Current time: 2019-03-05 16:15:01.627130179 +0100 CET m=+17.001731232; waiting for ticker
Current time: 2019-03-05 16:15:01.627210074 +0100 CET m=+17.001810973; Obtained time: 2019-03-05 16:14:58.625636409 +0100 CET m=+14.000237291
```
As you can see, when the time between ticks is greater than the ticker's interval, the loop does not have to wait for a tick... Taking this to our command, the time between iterations will be 10 minutes minus whatever the chef-run lasts, with a minimum of zero, instead of a minimum of 10 minutes.

Reply:
Done